File: ios/RNSPUserData/RNSPUserData.swift
//
//  RNSPUserData.swift
//  sourcepoint-react-native-cmp
//
//  Created by Andre Herculano on 21/5/24.
//

import Foundation
import ConsentViewController

struct RNSPUserData: Encodable {
    let gdpr: RNSPGDPRConsent?
    let usnat: RNSPUSNatConsent?
    let preferences: RNSPPreferencesConsent?
    let globalcmp: RNSPGlobalCMPConsent?

    init(
        gdpr: RNSPGDPRConsent? = nil,
        usnat: RNSPUSNatConsent? = nil,
        preferences: RNSPPreferencesConsent? = nil,
        globalcmp: RNSPGlobalCMPConsent? = nil
    ) {
        self.gdpr = gdpr
        self.usnat = usnat
        self.preferences = preferences
        self.globalcmp = globalcmp
    }
}

extension RNSPUserData {
    init(from sdkConsents: SPUserData?) {
        guard let sdkConsents = sdkConsents else {
            self.init()
            return
        }

        self.init(
            gdpr: RNSPGDPRConsent(from: sdkConsents.gdpr?.consents),
            usnat: RNSPUSNatConsent(from: sdkConsents.usnat?.consents),
            preferences: RNSPPreferencesConsent(from: sdkConsents.preferences?.consents),
            globalcmp: RNSPGlobalCMPConsent(from: sdkConsents.globalcmp?.consents)
        )
    }
}

File: android/src/main/java/com/sourcepoint/reactnativecmp/consents/RNSPUserData.kt
package com.sourcepoint.reactnativecmp.consents

import com.facebook.react.bridge.Arguments.createMap
import com.facebook.react.bridge.ReadableMap
import com.sourcepoint.cmplibrary.model.exposed.SPConsents

interface RNMappable {
  fun toRN(): ReadableMap
}

data class RNSPUserData(
  val gdpr: RNSPGDPRConsent?,
  val usnat: RNSPUSNatConsent?,
  val preferences: RNSPPreferencesConsent?,
  val globalCMP: RNSPGlobalCMPConsent?
): RNMappable {
  constructor(spData: SPConsents): this(
    gdpr = spData.gdpr?.let { RNSPGDPRConsent(gdpr = it.consent)},
    usnat = spData.usNat?.let { RNSPUSNatConsent(usnat = it.consent)},
    preferences = spData.preferences?.let { RNSPPreferencesConsent(preferences = it.consent) },
    globalCMP = spData.globalcmp?.let { RNSPGlobalCMPConsent(globalCMP = it.consent) }
  )

  override fun toRN(): ReadableMap = createMap().apply {
    gdpr?.let { putMap("gdpr", it.toRN()) }
    usnat?.let { putMap("usnat", it.toRN()) }
    preferences?.let { putMap("preferences", it.toRN()) }
    globalCMP?.let { putMap("globalcmp", it.toRN()) }
  }
}
